### ISSUE TITLE
Duplicate record errors in workflows

### ISSUE THREAD

**Issue by jpiquot** [OrchardCMS/OrchardCore#3436]
Hi,

We still have sometimes "database is locked" errors. When this happens, duplicate key errors are blocking Worflows.

Restarting OC does not clear the issue. We have to delete the duplicate entry in WorkflowTypeStartActivitiesIndex.

```
2019-04-05 09:51:47.6767|Default|0HLLPJE52JTCO:00000001||Microsoft.AspNetCore.Server.Kestrel|ERROR|Connection id "0HLLPJE52JTCO", Request id "0HLLPJE52JTCO:00000001": An unhandled exception was thrown by the application. System.ArgumentException: An item with the same key has already been added. Key: 4jpqrrzbq2cnp08g132j6sxecb
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](List`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at OrchardCore.Workflows.Http.Startup.Configure(IApplicationBuilder app, IRouteBuilder routes, IServiceProvider serviceProvider) in C:\projects\orchard2\src\OrchardCore.Modules\OrchardCore.Workflows\Http\Startup.cs:line 76
   at OrchardCore.Modules.ModularTenantRouterMiddleware.ConfigureTenantPipeline(IApplicationBuilder appBuilder, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 125
   at OrchardCore.Modules.ModularTenantRouterMiddleware.<>c__DisplayClass6_0.<BuildTenantPipeline>b__0(IApplicationBuilder builder) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 96
   at Microsoft.AspNetCore.Mvc.Internal.MiddlewareFilterBuilderStartupFilter.<>c__DisplayClass0_0.<Configure>g__MiddlewareFilterBuilder|0(IApplicationBuilder builder)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.BuildTenantPipeline(IServiceProvider rootServiceProvider, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 105
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 71
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 100
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)    at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](List`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at OrchardCore.Workflows.Http.Startup.Configure(IApplicationBuilder app, IRouteBuilder routes, IServiceProvider serviceProvider) in C:\projects\orchard2\src\OrchardCore.Modules\OrchardCore.Workflows\Http\Startup.cs:line 76
   at OrchardCore.Modules.ModularTenantRouterMiddleware.ConfigureTenantPipeline(IApplicationBuilder appBuilder, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 125
   at OrchardCore.Modules.ModularTenantRouterMiddleware.<>c__DisplayClass6_0.<BuildTenantPipeline>b__0(IApplicationBuilder builder) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 96
   at Microsoft.AspNetCore.Mvc.Internal.MiddlewareFilterBuilderStartupFilter.<>c__DisplayClass0_0.<Configure>g__MiddlewareFilterBuilder|0(IApplicationBuilder builder)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.BuildTenantPipeline(IServiceProvider rootServiceProvider, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 105
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 71
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 100
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)

```

```
2019-04-05 09:04:51.2594|Default|0HLLP12PLPULC:0000000D||Microsoft.AspNetCore.Server.Kestrel|ERROR|Connection id "0HLLP12PLPULC", Request id "0HLLP12PLPULC:0000000D": An unhandled exception was thrown by the application. Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 6: 'database table is locked'.
   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteScalar()
   at System.Data.Common.DbCommand.ExecuteScalarAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteScalarImplAsync[T](IDbConnection cnn, CommandDefinition command) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 1226
   at YesSql.Session.UpdateEntityAsync(Object entity)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.Dispose()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at OrchardCore.Hosting.ShellBuilders.ShellContext.ServiceScopeWrapper.Dispose() in C:\projects\orchard2\src\OrchardCore\OrchardCore.Abstractions\Shell\Builders\ShellContext.cs:line 279
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 118
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)    at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteScalar()
   at System.Data.Common.DbCommand.ExecuteScalarAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteScalarImplAsync[T](IDbConnection cnn, CommandDefinition command) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 1226
   at YesSql.Session.UpdateEntityAsync(Object entity)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.Dispose()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at OrchardCore.Hosting.ShellBuilders.ShellContext.ServiceScopeWrapper.Dispose() in C:\projects\orchard2\src\OrchardCore\OrchardCore.Abstractions\Shell\Builders\ShellContext.cs:line 279
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 118
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)

```
![image](https://user-images.githubusercontent.com/6775094/55616423-11fac680-5792-11e9-8636-3b7399bf5d54.png)

![image](https://user-images.githubusercontent.com/6775094/55615491-e5de4600-578f-11e9-9d6b-799e69e02651.png)

**Comment by sebastienros** [comment]
Any hint on how I can repro?

**Comment by jpiquot** [comment]
Not yet, I don't have this very often. It's quite difficult to repro. I did not have any issues for a week now but we made a major change : ContentItems do not use YesSQL anymore as we use a content manager that uses Microsoft Orleans Actors as a persistence layer.

I think that the issue may be due to making CRUD operations inside Workflows as ContentItemManger and WorkflowManager share the same YesSql Session.

**Comment by jpiquot** [comment]
We have another weird thing with workflows : If I try to delete the workflow task below, I have a timeout error on the database connection but the database has no issues. 

![image](https://user-images.githubusercontent.com/6775094/56648537-ceee8d80-6683-11e9-9a58-cd1cf2f59de2.png)



An unhandled exception occurred while processing the request.
Win32Exception: The wait operation timed out
Unknown location

SqlException: Timeout expired. The timeout period elapsed prior to completion of the operation or the server is not responding.
System.Data.SqlClient.SqlConnection.OnError(SqlException exception, bool breakConnection, Action<Action> wrapCloseInAction)

Stack Query Cookies Headers
Win32Exception: The wait operation timed out

Show raw exception details
SqlException: Timeout expired. The timeout period elapsed prior to completion of the operation or the server is not responding.
System.Data.SqlClient.SqlConnection.OnError(SqlException exception, bool breakConnection, Action<Action> wrapCloseInAction)
System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, bool breakConnection, Action<Action> wrapCloseInAction)
System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, bool callerHasConnectionLock, bool asyncClose)
System.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
System.Data.SqlClient.SqlCommand.EndExecuteNonQuery(IAsyncResult asyncResult)
System.Threading.Tasks.TaskFactory<TResult>.FromAsyncCoreLogic(IAsyncResult iar, Func<IAsyncResult, TResult> endFunction, Action<IAsyncResult> endAction, Task<TResult> promise, bool requiresSynchronization)
Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, object param) in SqlMapper.Async.cs
YesSql.Session.FlushAsync()
YesSql.Session.Dispose()
Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
OrchardCore.Hosting.ShellBuilders.ShellContext+ServiceScopeWrapper.Dispose() in ShellContext.cs
OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in ModularTenantContainerMiddleware.cs
Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

Show raw exception details
System.Data.SqlClient.SqlException (0x80131904): Timeout expired.  The timeout period elapsed prior to completion of the operation or the server is not responding. ---> System.ComponentModel.Win32Exception (258): The wait operation timed out
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at System.Data.SqlClient.SqlCommand.EndExecuteNonQuery(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, Object param) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 678
   at YesSql.Session.FlushAsync()
   at YesSql.Session.Dispose()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at OrchardCore.Hosting.ShellBuilders.ShellContext.ServiceScopeWrapper.Dispose() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Shell\Builders\ShellContext.cs:line 279
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 117
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
ClientConnectionId:8fae59f1-4878-4da0-814f-d95fff8dbbb4
Error Number:-2,State:0,Class:11

**Comment by jpiquot** [comment]
I think that all these locking issues won't go away until YesSql stops using pessimistic locking. With optimistic concurrency checks these issues may go away.

![image](https://user-images.githubusercontent.com/6775094/56650847-2a227f00-6688-11e9-83b7-04d6b36f2fc9.png)

**Comment by jtkech** [comment]
Looking at the 1st comments.

- In the `WorkflowTypeStartActivitiesIndex` table we can have 2 records with the same `WorkflowTypeId`, `Name` and so on, but yes the issue is that you also have the same `StartActivityId`. But this is not related to any worflow execution, this table is only updated when edit / save a `WorkFlowType`, so it's another issue, maybe e.g a `double-click` issue.

- Because it seems that you don't have a duplicate `WorkFlowType` document, one suggestion would be to use an unique constraint for the `WorkflowTypeStartActivitiesIndex` table.

      SchemaBuilder.CreateMapIndexTable(nameof(WorkflowTypeStartActivitiesIndex), table => table
          .Column<string>("WorkflowTypeId")
          .Column<string>("Name")
          .Column<bool>("IsEnabled")
          .Column<string>("StartActivityId", c => c.Unique())
          .Column<string>("StartActivityName")
      );

- More globally, in `WorkflowManager.TriggerEventAsync()` we can start / resume multiple workflow instances, not concurrently but in the same scope context and session (request scope or background scope if related to a timer event). I think that here **each worflow instance would need to be excuted in its own isolated scope and session** that is committed before executing the next one. Otherwise this could lead to db locks if some items are created / updated.

- Note: I simulated a double entry in the `WorkflowTypeStartActivitiesIndex` table with the same `StartActivityId` related to a timer event. Indeed, when the timer expires 2 workflow instances (of the same type) are executed one after the other but in the same scope context.

- Finally, in a request scope the session is committed asynchronously by our middleware, but not if in a background scope, here the session is committed synchronously when the scope is disposed. And, if we do the above it would be the same for each individual scope created for each workflow instance. So, **would be better to always commit the session asynchronously** as we plan to do in a pending PR.

**Comment by sebastienros** [comment]
Can someone try to repro this issue?

**Comment by jtkech** [comment]
Regarding the comment i did above

> Note: I simulated a double entry in the WorkflowTypeStartActivitiesIndex table with the same StartActivityId related to a timer event. Indeed, when the timer expires 2 workflow instances (of the same type) are executed one after the other but in the same scope context.

Maybe due to the timer event issue, see my comment in #5830 where the timer event can create multiple halted instances that get unblocked at the same time => multiple instances of the same workflow type execute in the same scope.

Also, many changes have been done around yesSql.

So i marked this issue as fixed by #5830.
